howtos/module/_posts/2016-07-07-Common-mistakes-while-coding-module.md

```diff
@@ -0,0 +1,183 @@
+---
+layout: post
+title:  "10+ common mistakes while coding a PrestaShop module"
+subtitle: "Do you commit these mistakes? Do not make them further."
+date: 2016-07-20 17:10:00
+author: [ kpodemski ]
+icon: icon-bug
+tags: [ howto, tutorial, modules ]
+published: false
+---
+
+PrestaShop has been around for many years, during which thousands of modules were made by thousands of developers, yet we still see the same simple mistakes being made, even thought they could easily be avoided.
+
+Here is a list of 10+ common mistakes when coding a PrestaShop module, with a short information on how to avoid them.
+
+<div class="alert alert-note" role="alert">
+<h4><i class='icon-file'></i> Note</h4>
+Several tips only apply to PrestaShop 1.5-1.6 module development. [PrestaShop 1.7 changes a few things](http://build.prestashop.com/news/module-development-changes-in-17/)!
+</div>
+
+* TOC
+{:toc}
+
+
+### 1. Creating/using a module with hardcoded database prefix
+
+This can be surprising to some but, yes, I saw this a few times: developers using the "ps\_" prefix in their SQL schemas, or SQL queries. 
+
+How to avoid this? Always use PrestaShop's `_DB_PREFIX_` constant, which is the prefix that is defined by the user during installation.
+
+
+### 2. Making sure your module can be translated
+
+PrestaShop 1.6 has an amazing translations system. Users can easily translate themes and modules from their back office, but not all themes/modules are available there, why so? Some module developers still forget to make their creation translatable, or make mistakes by using not the proper code for a translations.
+
+Remember:
+
+- Translations inside module files:
+`$this->l('Translatable text')`
+
+- Translations inside the module's back-office controller (which extends ModuleAdminController):
+`$this->l('Translatable text')`
+
+- Translations inside the module's front-office controller (which extends ModuleFrontController):
+`$this->module->l('Translatable text', 'controller_name')`
+
+- Translations inside mails:
+`Mail::l('Translatable text')`
+
+- Translations inside PDF files:
+`PDF:l('Translatable text')`
+
+Remember to use English as your default language in translations! Indeed, it is always easier to find people to translate from English than other languages.
+
+Note that PrestaShop 1.7 completely [updates the way translation is defined](http://build.prestashop.com/news/new-translation-system-prestashop-17/#calling-symfonys-translator).
+
+
+### 3. Data validation - do not trust your users or yourself!
+
+Recent events show that even the most popular modules or themes can be affected security flow and exploited by hackers. There are a few simple steps which will help you keep your modules secure:
```

Comment:
Ohhww, it's a about a security flaw and not the security (work) flow. Haha, how a simple typo can turn things around :stuck_out_tongue_winking_eye: